R/fotmob_stats.R

```diff
@@ -60,35 +76,63 @@
     league_name = rlang::maybe_missing(league_name, NULL),
     league_id = rlang::maybe_missing(league_id, NULL)
   )
+
   url <- sprintf(
     "https://www.fotmob.com%s/%ss",
     stringr::str_replace(tables$page_url[1], "overview", "stats"),
     team_or_player
   )
   page <- url %>% rvest::read_html()
+
+  options <- page |> rvest::html_elements("option")
+  has_options <- length(options) > 0
+
   see_all_button <- page %>% rvest::html_elements(".SeeAllButton")
   has_see_all_button <- length(see_all_button) > 0
 
-  if(has_see_all_button) {
+  if (has_options) {
+
+    values <- options |> rvest::html_attr("value")
+
+    parts <- values |>
+      purrr::keep(~stringr::str_detect(.x, "season")) |>
+      stringr::str_split("/")
+
+    if(length(parts) == 0) {
+      rlang::abort(glue::glue("Could not parse season ids from {url}."))
+    }
+
+    if(length(parts[[1]]) != 5) {
+      rlang::abort(glue::glue("Season ids not stored in expected format at {url}."))
+    }
+
+    season_ids <- parts |>
+      purrr::map_chr(~purrr::pluck(.x, 4))
+
+    ## protect against the current season being in the offseason, unless there is no other season.
+    season_id <- ifelse(length(season_ids) > 1, season_ids[2], season_ids[1])
+
+    next_url <- sprintf(
+      "https://www.fotmob.com/leagues/%s/stats/season/%s/%ss/saves_team",
+      tables$league_id[1],
+      season_id,
+      team_or_player
+    )
+
+    next_page <- next_url |> rvest::read_html()
+    next_options <- next_page |> rvest::html_elements("option")
+    .extract_seasons_and_stats_from_options(next_options)
+
+  } else if (has_see_all_button) {
+
     hrefs <- see_all_button %>% rvest::html_attr("href")
     next_url <- sprintf(
       "https://www.fotmob.com%s",
       hrefs[1]
     )
     next_page <- next_url %>% rvest::read_html()
-    option_elements <- next_page %>% rvest::html_elements("option")
-    labels <- option_elements %>% rvest::html_text2()
-    rgx <- "(^.*)(\\s)(20[012].*$)"
-    league_name <- labels %>% stringr::str_replace(rgx, "\\1")
-    season_name <- labels %>% stringr::str_replace(rgx, "\\3")
-    league_name <- ifelse(season_name == league_name, NA_character_, league_name)
-    is_season <- season_name %>% stringr::str_detect("^2")
-    tibble::tibble(
-      league_name = league_name,
-      option_type = ifelse(is_season, "season", "stat"),
-      name = season_name,
-      id = option_elements %>% rvest::html_attr("value")
-    )
+    options <- next_page %>% rvest::html_elements("option")
+    .extract_seasons_and_stats_from_options(options)
```

Comment:
tbh i'm not sure if this "See All" button logic is actually needed anymore, but i can't say for sure until seasons start. This older logic depends on stats existing on the default stats page (e.g. https://www.fotmob.com/leagues/47/stats/premier-league/teams), but no stats are loaded right now when looking forward to the 2022/23 season (fotmob default right now).

The logic that i've added in this PR extracts season ids from select options that are hidden. For example, it looks like this

```
{xml_nodeset (7)}
[1] <option value="stats/47/season/17664/topstats.json">Season 2022/2023<!-- -->\n</option>\n
[2] <option value="stats/47/season/16390/topstats.json">Season 2021/2022<!-- -->\n</option>\n
[3] <option value="stats/47/season/15382/topstats.json">Season 2020/2021<!-- -->\n</option>\n
[4] <option value="stats/47/season/14022/topstats.json">Season 2019/2020<!-- -->\n</option>\n
[5] <option value="stats/47/season/12776/topstats.json">Season 2018/2019<!-- -->\n</option>\n
[6] <option value="stats/47/season/11522/topstats.json">Season 2017/2018<!-- -->\n</option>\n
[7] <option value="stats/47/season/10418/topstats.json">Season 2016/2017<!-- -->\n</option>
```'

It then guesses that the "Team saves" stat page exists in the prior season for a given league, e.g. at https://www.fotmob.com/leagues/47/stats/season/16390/teams/saves_team

Finally, it uses existing logic for extracting stat names (e.g. things that look like `saves_team`).

Reply:
Nice - better to keep it in for now and we can always reassess when the seasons start